File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <unordered_set>
using namespace std;


int solution(vector<int> v) {
    int answer = 0;

    int arr[10001] = {0,};
    unordered_set<int> s;
    for (int i = 0; i < v.size(); i++) {
        s.insert(v[i]);
        arr[v[i]]++;
    }
    unordered_set<int> a;
    for (int i = 0; i < v.size(); i++) {
        if (arr[v[i]] >= 2) {
            arr[v[i]]--;
        } else {
            arr[v[i]]--;
            s.erase(v[i]);
        }
        a.insert(v[i]);

        if (a.size() == s.size())
            answer++;
    }

    return answer;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = -1;
        
        Map<Integer, Integer> map = new HashMap<>();
        Set<Integer> set = new HashSet<>();
        
        for (int t : topping) {
            map.put(t, map.getOrDefault(t, 0) + 1);
        }

        for (int i = 0; i < topping.length - 1; i++) {
            int t = topping[i];
            
            int cnt = map.get(t);
            
            if (cnt > 0) {
                set.add(t);
                if (cnt - 1 == 0) {
                    map.remove(t);
                } else {
                    map.put(t, map.get(t) - 1);
                }
            }
            
            int myCnt = map.size();
            int broCnt = set.size();
            
            if (myCnt == broCnt) answer++;
        }
        
        return answer + 1;
    }
}